bootloader/src/pyi_launch.c

```diff
@@ -338,18 +357,27 @@ int pyi_launch_run_scripts(ARCHIVE_STATUS *status)
 			strcat(buf, ".py");
 			VS("LOADER: Running %s\n", buf);
             if (is_py2) {
-	      __file__ = PI_PyString_FromString(buf);
-	    } else {
-	      __file__ = PI_PyUnicode_FromString(buf);
-	    };
+              __file__ = PI_PyString_FromString(buf);
+            } else {
+              __file__ = PI_PyUnicode_FromString(buf);
+            };
             PI_PyObject_SetAttrString(__main__, "__file__", __file__);
             Py_DECREF(__file__);
+
+            /* Unmarshall code object */
+            code = PI_PyMarshal_ReadObjectFromString(data, ntohl(ptoc->ulen));
+            if(!code) {
+                FATALERROR("Failed to unmarshal code object for %s\n", ptoc->name);
+			    PI_PyErr_Print();
+                return -1;
+            }
 			/* Run it */
-			rc = PI_PyRun_SimpleString((char *) data);
-			/* log errors and abort */
-			if (rc != 0) {
-				FATALERROR("%s returned %d\n", ptoc->name, rc);
-				return rc;
+			retval = PI_PyEval_EvalCode(code, main_dict, main_dict);
+			/* If retval is NULL, an error occured. Otherwise, it is a Python object */
+			if (!retval) {
+                FATALERROR("Failed to execute script %s\n", ptoc->name);
+                PI_PyErr_Print();
+				return -1;
```

Comment:
I assume we need to `DECREF(retavl)` here. At east this is what py2exe, cx_freeze and [PyRun_InteractiveOneObject](https://hg.python.org/releases/2.7.11/file/tip/Python/pythonrun.c#l240) do.